File: packages/posthog/ios/Plugin/Classes/Options/CaptureExceptionOptions.swift
import Foundation
import Capacitor

@objc public class CaptureExceptionOptions: NSObject {
    private var message: String
    private var name: String?
    private var stacktrace: [StackFrame]?
    private var properties: [String: Any]?

    init(message: String, name: String?, stacktrace: JSArray?, properties: JSObject?) {
        self.message = message
        self.name = name
        self.stacktrace = CaptureExceptionOptions.createStackFrames(stacktrace)
        self.properties = PosthogHelper.createHashMapFromJSObject(properties)
    }

    func getMessage() -> String {
        return message
    }

    func getName() -> String? {
        return name
    }

    func getStacktrace() -> [StackFrame]? {
        return stacktrace
    }

    func getProperties() -> [String: Any]? {
        return properties
    }

    private static func createStackFrames(_ stacktrace: JSArray?) -> [StackFrame]? {
        guard let stacktrace = stacktrace else {
            return nil
        }
        var frames: [StackFrame] = []
        for element in stacktrace {
            if let frame = element as? JSObject {
                frames.append(StackFrame(frame))
            }
        }
        return frames
    }
}

File: packages/posthog/android/src/main/java/io/capawesome/capacitorjs/plugins/posthog/classes/options/CaptureExceptionOptions.java
package io.capawesome.capacitorjs.plugins.posthog.classes.options;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import io.capawesome.capacitorjs.plugins.posthog.PosthogHelper;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.json.JSONException;
import org.json.JSONObject;

public class CaptureExceptionOptions {

    @NonNull
    private final String message;

    @Nullable
    private final String name;

    @Nullable
    private final List<StackFrame> stacktrace;

    @Nullable
    private final Map<String, Object> properties;

    public CaptureExceptionOptions(
        @NonNull String message,
        @Nullable String name,
        @Nullable JSArray stacktrace,
        @Nullable JSObject properties
    ) throws JSONException {
        this.message = message;
        this.name = name;
        this.stacktrace = createStackFrames(stacktrace);
        this.properties = PosthogHelper.createHashMapFromJSONObject(properties);
    }

    @NonNull
    public String getMessage() {
        return message;
    }

    @Nullable
    public String getName() {
        return name;
    }

    @Nullable
    public List<StackFrame> getStacktrace() {
        return stacktrace;
    }

    @Nullable
    public Map<String, Object> getProperties() {
        return properties;
    }

    @Nullable
    private static List<StackFrame> createStackFrames(@Nullable JSArray stacktrace) throws JSONException {
        if (stacktrace == null) {
            return null;
        }
        List<StackFrame> frames = new ArrayList<>();
        for (int i = 0; i < stacktrace.length(); i++) {
            JSONObject frame = stacktrace.getJSONObject(i);
            frames.add(new StackFrame(frame));
        }
        return frames;
    }
}
